### PR TITLE
[system-probe] Add packet counting to existing UDP byte counters

### PR DESCRIPTION
**Pull request by pimlu**
### What does this PR do?

This PR begins tracing UDP packet counts in the agent and adds some tests to ensure it works as expected for both IPv4 and IPv6.

### Motivation

Context - [NPM-2973](https://datadoghq.atlassian.net/browse/NPM-2973). The agent already has tracked UDP byte counts so it was determined we can leverage the existing kprobes to track packet counts in the same places.

### Possible Drawbacks / Trade-offs

This does not track UDP sends that have corking or MSG_MORE which are not commonly used. Doing this requires tracing `udp_send_skb` which probably will have to wait until after we deprecate prebuilt as mentioned in [this thread](https://github.com/DataDog/datadog-agent/pull/28546#discussion_r1729110795)

### Describe how to test/QA your changes

This PR leverages the existing test suite and adds some new ones:
```
DD_SYSTEM_PROBE_CONFIG_BPF_DEBUG=true sudo -E go test -v -tags "linux linux_bpf npm test" ./pkg/network/tracer --run TestTracerSuite
```

[NPM-2973]: https://datadoghq.atlassian.net/browse/NPM-2973?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by akarpz** [comment]
can you add a description?

File: pkg/network/tracer/tracer_linux_test.go [hunk 1]
```diff
@@ -1195,12 +1195,99 @@ func (s *TracerSuite) TestUDPPeekCount() {
...
+		for i := 0; i < COUNT; i++ {
+			flags := 0
+			isLast := i == COUNT-1
+			if !isLast {
+				flags |= syscall.MSG_MORE
+			}
+			sockAddr := syscall.SockaddrInet4{Port: raddr.Port}
+			copy(sockAddr.Addr[:], raddr.IP.To4())
+
+			t.Log("flags", flags)
```
[the remark this answers is not in the document]
Comment (pimlu) [reply pkg/network/tracer/tracer_linux_test.go]: Actually removed this test of corking in general.

Reason: our existing probes don't cover corking and it's a kind of obscure UDP feature. Supporting corking more or less requires tracing `udp_send_skb` (which increments the global SNMP counter `UDP_MIB_OUTDATAGRAMS`) but its arguments only have the skb, not the sk. That means the prebuilt probe would have to offset guess `skb->sk` which Hasan says is not reliable on older kernel versions.

So, might make sense to revisit and trace `udp_send_skb` once we deprecate prebuilt

File: pkg/network/tracer/tracer_linux_test.go [hunk 2]
```diff
@@ -1203,12 +1247,81 @@ func (s *TracerSuite) TestUDPPeekCount() {
...
+	tr := setupTracer(t, config)
+
+	ln, c, err := setupUdpSockets(udpnet, ip)
+	require.NoError(t, err)
+	defer ln.Close()
+	defer c.Close()
+
+	msg := []byte("asdf")
+	// send UDP packets of increasing length
+	for i := range len(msg) {
```
Comment (hmahmood) [inline pkg/network/tracer/tracer_linux_test.go:1286]: Just 
```go
for i := range msg {
```
would give you the indices.

File: pkg/network/tracer/tracer_linux_test.go [hunk 3]
```diff
@@ -1126,31 +1126,75 @@ func (s *TracerSuite) TestSelfConnect() {
...
+func setupUdpSockets(udpnet, ip string) (*net.UDPConn, *net.UDPConn, error) {
+	serverAddr := fmt.Sprintf("%s:%d", ip, 0)
+
+	laddr, err := net.ResolveUDPAddr(udpnet, serverAddr)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var ln, c *net.UDPConn = nil, nil
+	defer func() {
```
Comment (hmahmood) [inline pkg/network/tracer/tracer_linux_test.go:1140]: `t.Cleanup` would be preferable to `defer` here. You can pass in `t *testing.T` as a parameter to the function.

Comment (pimlu) [reply pkg/network/tracer/tracer_linux_test.go]: Thanks, updated

File: pkg/network/tracer/tracer_linux_test.go [hunk 4]
```diff
@@ -1506,10 +1606,18 @@ func (s *TracerSuite) TestSendfileRegression() {
 		}, 3*time.Second, 100*time.Millisecond, "couldn't find connections used by sendfile(2)")
 
 		if assert.NotNil(t, outConn, "couldn't find outgoing connection used by sendfile(2)") {
-			assert.Equalf(t, int64(clientMessageSize), int64(outConn.Monotonic.SentBytes), "sendfile send data wasn't properly traced")
+			assert.Equalf(t, int64(clientMessageSize), int64(outConn.Monotonic.SentBytes), "sendfile sent bytes wasn't properly traced")
+			if connType == network.UDP {
```
Comment (hmahmood) [inline pkg/network/tracer/tracer_linux_test.go:1610]: Are we expecting anything other than UDP here?

Comment (hmahmood) [reply pkg/network/tracer/tracer_linux_test.go:1610]: Nvm, this is shared code.

**Review by hestonhoffman — APPROVED** [review]
Docs 👍

**Review by hmahmood — APPROVED** [review]

**Pull request merged**